internal/db/admin.go

```diff
@@ -0,0 +1,668 @@
+	if !adminInfo.AdminDesc.Policy.HasValue() {
+		return ErrAACIsEnabledButIsMissingPolicyInfo
+	}
+
+	policyID := adminInfo.AdminDesc.Policy.Value().ID
+	resourceName := adminInfo.AdminDesc.Policy.Value().ResourceName
+	if policyID == "" || resourceName == "" {
+		return ErrAACIsEnabledButIsMissingPolicyInfo
+	}
+	// Since public node will have unrestricted access, the object we are gating MUST be registered
+	// if admin access control is configured.
+	isRegistered, err := adminInfo.AdminACP.ObjectOwner(
+		ctx,
+		policyID,
+		resourceName,
+		objectID,
+	)
+	if err != nil {
+		return err
+	}
+
+	if !isRegistered.HasValue() {
+		return ErrAACNodeObjectToGateIsNotRegistered
+	}
+
+	var identityValue string
+	if !identity.HasValue() {
+		// We can't assume that there is no-access just because there is no identity even if the operation
+		// is registered with acp, this is because it is possible that acp has a registered relation targeting
+		// "*" (any) actor which would mean that even a request without an identity might be able to access
+		// an operation registered with acp. So we pass an empty `did` to accommodate that case.
+		identityValue = ""
+	} else {
+		identityValue = identity.Value().DID()
+	}
+
+	adminResourcePerm, ok := permission.(acpTypes.AdminResourcePermission)
+	if !ok {
+		return client.ErrInvalidResourcePermissionType
+	}
+
+	// Now actually check using the signature if this identity has access or not.
+	hasAccess, err := adminInfo.AdminACP.VerifyAccessRequest(
+		ctx,
+		adminResourcePerm,
+		identityValue,
+		policyID,
+		resourceName,
+		objectID,
+	)
+
+	if err != nil {
+		return acp.NewErrFailedToVerifyAdminAccessWithACP(
+			err,
+			"AdminACP",
+			permission.String(),
+			policyID,
+			identityValue,
+			resourceName,
+			objectID,
+		)
+	}
+
+	if hasAccess {
+		return nil
+	}
+
+	return client.ErrNotAuthorizedToPerformOperation
+}
+
+func (db *DB) addAACActorRelationship(
+	ctx context.Context,
+	relation string,
+	targetActor string,
+) (client.AddActorRelationshipResult, error) {
+	// Note: While we give unrestricted access when admin acp if turned off, there are certain
+	// requests that we can't do when admin acp is turned off or unavailable, this is one of them.
+	if !db.adminInfo.AdminDesc.IsEnabled || db.adminInfo.AdminACP == nil {
+		return client.AddActorRelationshipResult{}, client.ErrACPOperationButACPNotAvailable
+	}
+
+	if !db.adminInfo.AdminDesc.Policy.HasValue() {
+		return client.AddActorRelationshipResult{}, ErrAACIsEnabledButIsMissingPolicyInfo
+	}
+
+	policyDesc := db.adminInfo.AdminDesc.Policy.Value()
+	if policyDesc.ID == "" || policyDesc.ResourceName == "" {
+		return client.AddActorRelationshipResult{}, ErrAACIsEnabledButIsMissingPolicyInfo
+	}
+
+	requestActor := identity.FromContext(ctx)
+	if !requestActor.HasValue() || requestActor.Value() == nil || requestActor.Value().DID() == "" {
+		return client.AddActorRelationshipResult{}, ErrAACRelationshipOperationRequiresIdentity
+	}
```

Comment:
thought: These checks seem to be necessary in a few locations. Might be nice to have a function that takes care of it.

Reply:
I think seeing all the steps might make sense in test definitions but I don't think it applies here. Not a blocker for me but I still think it would be nicer to remove the duplication 🙂